File: scraper.js
const axios = require('axios');
const cheerio = require('cheerio');
const mysql = require('mysql2/promise');
const fs = require('fs').promises;
const crypto = require('crypto');
require('dotenv').config();

// Configuration
const config = {
    db: {
        host: process.env.DB_HOST || 'localhost',
        user: process.env.DB_USER || 'root',
        password: process.env.DB_PASSWORD || '',
        database: process.env.DB_NAME || 'vibe_coach'
    },
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
    delayBetweenRequests: 2000, // 2 seconds
    maxRetries: 3
};

class FlirtingLineScraper {
    constructor() {
        this.db = null;
        this.scrapedCount = 0;
        this.duplicateCount = 0;
        this.lines = [];
    }

    async connect() {
        try {
            this.db = await mysql.createConnection(config.db);
            console.log('✓ Connected to database\n');
            return true;
        } catch (error) {
            console.error('❌ Database connection failed:', error.message);
            return false;
        }
    }

    async close() {
        if (this.db) {
            await this.db.end();
        }
    }

    hashLine(text) {
        return crypto.createHash('sha256').update(text.toLowerCase().trim()).digest('hex');
    }

    async lineExists(text) {
        const hash = this.hashLine(text);
        const [rows] = await this.db.query(
            'SELECT COUNT(*) as count FROM flirting_lines WHERE SHA2(LOWER(TRIM(line_text)), 256) = ?',
            [hash]
        );
        return rows[0].count > 0;
    }

    cleanText(text) {
        return text
            .replace(/\s+/g, ' ')
            .trim()
            .replace(/^["'""'']+|["'""'']+$/g, '');
    }

    async insertLine(lineData) {
        const { text, category, style, context, quality_score, source_type } = lineData;

        if (await this.lineExists(text)) {
            this.duplicateCount++;
            return false;
        }

        try {
            const query = `
                INSERT INTO flirting_lines 
                (line_text, category, style, context, stage, target_gender, source_type, quality_score, originality_score, success_rate, usage_count)
                VALUES (?, ?, ?, ?, 'initial', 'neutral', ?, ?, 3.5, 0.5, 0)
            `;
            await this.db.query(query, [
                text,
                category || 'playful',
                style || 'international',
                context || 'opening',
                source_type || 'website',
                quality_score || 3.0
            ]);
            this.scrapedCount++;
            this.lines.push(lineData);
            return true;
        } catch (error) {
            console.error('⚠️  Error inserting line:', error.message);
            return false;
        }
    }

    async fetchWithRetry(url, retries = config.maxRetries) {
        for (let i = 0; i < retries; i++) {
            try {
                const response = await axios.get(url, {
                    headers: { 'User-Agent': config.userAgent },
                    timeout: 10000
                });
                return response.data;
            } catch (error) {
                if (i === retries - 1) throw error;
                await this.delay(1000 * (i + 1));
            }
        }
    }

    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async scrapePickupLinesOrg() {
        console.log('\n📝 Scraping pickup lines from web sources...');

        const sources = [
            {
                url: 'https://www.pickuplinesgalore.com/funny.html',
                category: 'funny',
                style: 'international'
            },
            {
                url: 'https://www.pickuplinesgalore.com/cheesy.html',
                category: 'cheesy',
                style: 'international'
            }
        ];

        for (const source of sources) {
            try {
                console.log(`  Fetching ${source.category} lines...`);
                const html = await this.fetchWithRetry(source.url);
                const $ = cheerio.load(html);

                // Find list items and paragraphs
                $('li, p').each((i, elem) => {
                    const text = this.cleanText($(elem).text());

                    // Filter: reasonable length and looks like a pickup line
                    if (text.length >= 10 && text.length <= 200) {
                        this.insertLine({
                            text,
                            category: source.category,
                            style: source.style,
                            context: 'opening',
                            quality_score: 3.5,
                            source_type: 'website'
                        });
                    }
                });

                await this.delay(config.delayBetweenRequests);
            } catch (error) {
                console.error(`  ⚠️  Failed to scrape ${source.category}:`, error.message);
            }
        }
    }

    async addCuratedLines() {
        console.log('\n📝 Adding curated flirting lines...');

        const curatedLines = [
            { text: "Are you a magician? Because whenever I look at you, everyone else disappears.", category: "romantic", style: "international", quality_score: 4.7 },
            { text: "Do you have a map? I keep getting lost in your eyes.", category: "cheesy", style: "international", quality_score: 3.2 },
            { text: "Is your name Google? Because you have everything I've been searching for.", category: "clever", style: "international", quality_score: 3.8 },
            { text: "If beauty were time, you'd be eternity.", category: "romantic", style: "international", quality_score: 4.8 },
            { text: "Are you a time traveler? Because I see you in my future.", category: "romantic", style: "international", quality_score: 4.5 },
            { text: "Do you believe in love at first sight, or should I walk by again?", category: "playful", style: "international", quality_score: 4.3 },
            { text: "Are you made of copper and tellurium? Because you're Cu-Te.", category: "clever", style: "international", quality_score: 4.1 },
            { text: "I'm not a photographer, but I can picture us together.", category: "smooth", style: "international", quality_score: 4.2 },
            { text: "Do you have a Band-Aid? I just scraped my knee falling for you.", category: "cheesy", style: "international", quality_score: 3.3 },
            { text: "Is your name Wi-fi? Because I'm really feeling a connection.", category: "clever", style: "international", quality_score: 3.6 },
            { text: "You must be tired because you've been running through my mind all day.", category: "playful", style: "international", quality_score: 3.5 },
            { text: "If you were a vegetable, you'd be a cutecumber.", category: "funny", style: "international", quality_score: 3.7 },
            { text: "Are you French? Because Eiffel for you.", category: "clever", style: "international", quality_score: 4.2 },
            { text: "Do you like Star Wars? Because Yoda one for me!", category: "funny", style: "international", quality_score: 3.9 },
            { text: "I was wondering if you had an extra heart, because mine was just stolen.", category: "romantic", style: "international", quality_score: 4.6 },
            { text: "Are you a parking ticket? Because you've got FINE written all over you.", category: "cheesy", style: "international", quality_score: 3.5 },
            { text: "I don't have a pickup line. I just wanted to say you're stunning.", category: "direct", style: "international", quality_score: 4.7 },
            { text: "You're like a dream I never want to wake up from.", category: "romantic", style: "international", quality_score: 4.9 },
            { text: "I don't need to see the stars when I can look at your eyes.", category: "romantic", style: "international", quality_score: 4.8 },
            { text: "Are you a 90 degree angle? Because you're looking right.", category: "funny", style: "international", quality_score: 4.0 }
        ];

        for (const line of curatedLines) {
            await this.insertLine({
                ...line,
                context: 'opening',
                source_type: 'manual_entry'
            });
        }
    }

    async exportToCSV() {
        console.log('\n💾 Exporting to CSV...');

        const csvHeader = 'line_text,category,style,context,quality_score,source_type\n';
        const csvRows = this.lines.map(line =>
            `"${line.text}",${line.category},${line.style},${line.context},${line.quality_score},${line.source_type}`
        ).join('\n');

        const csvContent = csvHeader + csvRows;
        await fs.writeFile('scraped_lines.csv', csvContent);
        console.log(`✓ Exported ${this.lines.length} lines to scraped_lines.csv`);
    }

    async exportToSQL() {
        console.log('💾 Exporting to SQL...');

        let sqlContent = `-- Scraped Flirting Lines (${new Date().toISOString()})\n\n`;
        sqlContent += `USE vibe_coach;\n\n`;
        sqlContent += `INSERT INTO flirting_lines (line_text, category, style, context, stage, target_gender, source_type, quality_score, originality_score, success_rate, usage_count) VALUES\n`;

        const values = this.lines.map((line, i) => {
            const isLast = i === this.lines.length - 1;
            return `('${line.text.replace(/'/g, "''")}', '${line.category}', '${line.style}', '${line.context}', 'initial', 'neutral', '${line.source_type}', ${line.quality_score}, 3.5, 0.5, 0)${isLast ? ';' : ','}`;
        }).join('\n');

        sqlContent += values;

        await fs.writeFile('scraped_lines.sql', sqlContent);
        console.log(`✓ Exported ${this.lines.length} lines to scraped_lines.sql`);
    }

    async run() {
        console.log('='.repeat(60));
        console.log('🔍 Flirting Lines Web Scraper (Node.js)');
        console.log('='.repeat(60));

        const connected = await this.connect();
        if (!connected) {
            console.log('\n⚠️  Running without database - will export to files only\n');
        }

        try {
            // Add curated lines (always works)
            await this.addCuratedLines();

            // Try web scraping (may fail)
            console.log('\n⚠️  Note: Web scraping may fail due to site changes or network issues');
            try {
                await this.scrapePickupLinesOrg();
            } catch (error) {
                console.error('⚠️  Web scraping failed:', error.message);
            }

            // Export to files
            if (this.lines.length > 0) {
                await this.exportToCSV();
                await this.exportToSQL();
            }

            console.log('\n' + '='.repeat(60));
            console.log('✅ Scraping complete!');
            console.log(`   • New lines added to DB: ${this.scrapedCount}`);
            console.log(`   • Duplicates skipped: ${this.duplicateCount}`);
            console.log(`   • Total lines collected: ${this.lines.length}`);
            console.log('='.repeat(60));

        } finally {
            await this.close();
        }
    }
}

// Run the scraper
if (require.main === module) {
    const scraper = new FlirtingLineScraper();
    scraper.run().catch(console.error);
}

module.exports = FlirtingLineScraper;

File: scraper.py
import os
import sys
import time

# Fix encoding for Windows console
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')

import hashlib
import mysql.connector
from dotenv import load_dotenv
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
import re

# Load environment variables
load_dotenv()

# Database configuration
DB_CONFIG = {
    'host': os.getenv('DB_HOST', 'localhost'),
    'user': os.getenv('DB_USER', 'root'),
    'password': os.getenv('DB_PASSWORD', ''),
    'database': os.getenv('DB_NAME', 'vibe_coach')
}

# Scraping configuration
USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
DELAY_BETWEEN_REQUESTS = 2  # seconds
MAX_RETRIES = 3

class FlirtyLineScraper:
    def __init__(self):
        self.db = None
        self.cursor = None
        self.session = requests.Session()
        self.session.headers.update({'User-Agent': USER_AGENT})
        self.scraped_count = 0
        self.duplicate_count = 0
        
    def connect_db(self):
        """Connect to MySQL database"""
        try:
            self.db = mysql.connector.connect(**DB_CONFIG)
            self.cursor = self.db.cursor()
            print("✓ Connected to database\n")
        except mysql.connector.Error as e:
            print(f"❌ Database connection failed: {e}")
            return False
        return True
    
    def close_db(self):
        """Close database connection"""
        if self.cursor:
            self.cursor.close()
        if self.db:
            self.db.close()
    
    def line_exists(self, line_text):
        """Check if line already exists in database"""
        line_hash = hashlib.sha256(line_text.strip().lower().encode()).hexdigest()
        query = "SELECT COUNT(*) FROM flirting_lines WHERE SHA2(LOWER(TRIM(line_text)), 256) = %s"
        self.cursor.execute(query, (line_hash,))
        count = self.cursor.fetchone()[0]
        return count > 0
    
    def insert_line(self, line_text, category='General', style='International', source='Web Scraping'):
        """Insert a flirting line into database"""
        if self.line_exists(line_text):
            self.duplicate_count += 1
            return False
        
        try:
            query = """
                INSERT INTO flirting_lines (line_text, category, style, usage_count)
                VALUES (%s, %s, %s, 0)
            """
            self.cursor.execute(query, (line_text, category, style))
            self.db.commit()
            self.scraped_count += 1
            return True
        except mysql.connector.Error as e:
            print(f"⚠️  Error inserting line: {e}")
            return False
    
    def clean_text(self, text):
        """Clean and normalize text"""
        # Remove extra whitespace
        text = re.sub(r'\s+', ' ', text).strip()
        # Remove quotes if they wrap the entire text
        text = text.strip('"\'""''')
        return text
    
    def scrape_pickuplines_com(self):
        """Scrape pickup lines from pickuplines.com"""
        print("📝 Scraping pickuplines.com...")
        base_url = "https://www.pickuplines.com"
        categories = ['funny', 'cheesy', 'cute', 'romantic', 'corny']
        
        for cat in categories:
            try:
                url = f"{base_url}/{cat}-pickup-lines/"
                print(f"  Fetching {cat} lines...")
                
                response = self.session.get(url, timeout=10)
                if response.status_code != 200:
                    print(f"  ⚠️  Failed to fetch {url}")
                    continue
                
                soup = BeautifulSoup(response.content, 'lxml')
                
                # Find all pickup lines (adjust selector based on site structure)
                lines = soup.find_all(['p', 'li', 'div'], class_=re.compile(r'line|pickup|quote'))
                
                if not lines:
                    # Fallback: get all paragraphs
                    lines = soup.find_all('p')
                
                for line in lines:
                    text = self.clean_text(line.get_text())
                    
                    # Filter: must be between 10 and 200 characters
                    if 10 <= len(text) <= 200 and '?' not in text[:50]:
                        category_map = {
                            'funny': 'Witty',
                            'cheesy': 'Playful',
                            'cute': 'Sweet',
                            'romantic': 'Romantic',
                            'corny': 'Playful'
                        }
                        self.insert_line(text, category=category_map.get(cat, 'General'))
                
                time.sleep(DELAY_BETWEEN_REQUESTS)
                
            except Exception as e:
                print(f"  ⚠️  Error scraping {cat}: {e}")
    
    def scrape_thought_catalog(self):
        """Scrape pickup lines from Thought Catalog articles"""
        print("\n📝 Scraping Thought Catalog...")
        urls = [
            "https://thoughtcatalog.com/january-nelson/2021/05/pickup-lines/",
            "https://thoughtcatalog.com/january-nelson/2020/02/funny-pickup-lines/"
        ]
        
        for url in urls:
            try:
                print(f"  Fetching {url[:50]}...")
                response = self.session.get(url, timeout=10)
                
                if response.status_code != 200:
                    continue
                
                soup = BeautifulSoup(response.content, 'lxml')
                
                # Find list items and paragraphs
                items = soup.find_all(['li', 'p'])
                
                for item in items:
                    text = self.clean_text(item.get_text())
                    
                    # Must be a reasonable length and not a header/navigation
                    if 15 <= len(text) <= 200 and not text.startswith(('Related', 'Share', 'Tweet')):
                        self.insert_line(text, category='Witty', style='Modern')
                
                time.sleep(DELAY_BETWEEN_REQUESTS)
                
            except Exception as e:
                print(f"  ⚠️  Error: {e}")
    
    def scrape_manual_sources(self):
        """Add some manually curated lines as fallback"""
        print("\n📝 Adding curated lines...")
        
        curated_lines = [
            ("Are you a magician? Because whenever I look at you, everyone else disappears.", "Romantic", "Classic"),
            ("Do you have a map? I just got lost in your eyes.", "Sweet", "Classic"),
            ("Is your name Google? Because you have everything I've been searching for.", "Witty", "Modern"),
            ("Are you a parking ticket? Because you've got FINE written all over you.", "Playful", "Modern"),
            ("Do you believe in love at first sight, or should I walk by again?", "Playful", "Classic"),
            ("Are you a camera? Because every time I look at you, I smile.", "Sweet", "Modern"),
            ("If beauty were time, you'd be an eternity.", "Romantic", "Classic"),
            ("Are you made of copper and tellurium? Because you're Cu-Te.", "Witty", "Modern"),
            ("Do you have a Band-Aid? I just scraped my knee falling for you.", "Playful", "Classic"),
            ("Is your name Wi-Fi? Because I'm feeling a connection.", "Witty", "Modern"),
        ]
        
        for line, category, style in curated_lines:
            self.insert_line(line, category=category, style=style)
    
    def run(self):
        """Run the scraper"""
        print("=" * 60)
        print("🔍 Flirty Lines Web Scraper")
        print("=" * 60)
        
        if not self.connect_db():
            return
        
        try:
            # Add manual curated lines first (always works)
            self.scrape_manual_sources()
            
            # Try scraping websites (may fail due to network/structure)
            print("\n⚠️  Note: Web scraping may fail due to site changes or network issues")
            print("   The script will add what it can find.\n")
            
            try:
                self.scrape_pickuplines_com()
            except Exception as e:
                print(f"⚠️  Skipping pickuplines.com: {e}")
            
            try:
                self.scrape_thought_catalog()
            except Exception as e:
                print(f"⚠️  Skipping Thought Catalog: {e}")
            
            print("\n" + "=" * 60)
            print(f"✅ Scraping complete!")
            print(f"   • New lines added: {self.scraped_count}")
            print(f"   • Duplicates skipped: {self.duplicate_count}")
            print("=" * 60)
            
        finally:
            self.close_db()

if __name__ == "__main__":
    scraper = FlirtyLineScraper()
    scraper.run()
